velox/dwio/common/BitConcatenation.cpp

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/dwio/common/BitConcatenation.h"
+
+namespace facebook::velox::dwio::common {
+
+void BitConcatenation::append(
+    const uint64_t* FOLLY_NULLABLE bits,
+    int32_t begin,
+    int32_t end) {
+  int32_t numBits = end - begin;
+  if (!bits || bits::isAllSet(bits, begin, end, true)) {
+    appendOnes(numBits);
+    return;
+  }
+
+  ensureSpace(numBits_ + numBits);
```

Comment:
It seems it would be convenient for ensureSpace to return a uint64_t pointer to the buffer. This would allow to remove ->asMutable<uint64_t> call below and make code more readable.